### ISSUE TITLE
FengYun-4 LRIT/HRIT Support

### ISSUE THREAD

**Issue by Aang23** [altillimity/SatDump#167]
Support the FengYun-4x line of satellite downlinks :
- [x] LRIT
- [ ] HRIT-I
- [ ] HRIT-II (same freq as HRIT-III)
- [x] HRIT-III

Decoding process : 
- [x] DVB-S2 Demodulation
- [x] DVB-MPE UDP CADU Demux
- [ ] LRIT Files processing (+ DES decryption? Only if users have a valid key...)

LRIT is supposed to be fully encrypted. HRIT apparently contains some unencrypted data.

@sam210723 @Foxiks

**Comment by Aang23** [comment]
@Foxiks actually, on the same frequency as HRIT-III you obtained, there is HRIT-II. I'm not sure of the specifications, but maybe it has different modulation parameters? There should also be HRIT-I on another frequency.

**Comment by Foxiks** [comment]
unfortunately this satellite is at 3 degrees of elevation and gives a weak SNR for HRIT (excellent for LRIT). Also, this signal is not transmitted constantly and often turns off. I have been watching this. I did see a transmission on that frequency, but it was several times faster than the HRIT I gave you. But that fast transmission had an SNR of less than 1 dB.

**Comment by Foxiks** [comment]
I think @sam210723 could get HRIT with a good SNR :)

**Comment by sam210723** [comment]
I can give it a go, I've received LRIT before on a 90cm dish at about 27° elevation. Do we know what frequency HRIT-I is on?

![image](https://user-images.githubusercontent.com/5117093/198759947-4bd0a380-37a0-4ec5-a37f-d14c6d7eabad.png)

**Comment by Aang23** [comment]
Unfortunately, not yet. But I'd assume scanning around as we know LRIT/HRIT-II&III's freqs should do the trick if it isn't extremely weak.

**Comment by Foxiks** [comment]
@Aang23 I have a recording of an HRIT signal that is a 1-2 dB stronger than the one I sent you. Try with this: https://disk.yandex.ru/d/z5OozNZlHy_oxA

**Comment by Aang23** [comment]
![Screenshot from 2022-10-30 14-13-32](https://user-images.githubusercontent.com/23421917/198880343-f8c0e92e-7323-400e-8b52-63a4f5074732.png)

A bit better, but still noisy...

**Comment by Foxiks** [comment]
hm... Yes(

**Comment by Aang23** [comment]
Update : It turns out sometimes I'm a bit dumb :-)

So, LRIT is non-compliant, but a valid JPEG2000 bitstream is present, unencrypted in all the files I checked as of now... Just very offset compared to anything we should expect in xRIT.
![FY4A_9_20201208T161400Z](https://user-images.githubusercontent.com/23421917/201785255-0d9ae16f-af0f-49f9-a526-ae59a05f9e62.png)
![FY4A_12_20201208T161400Z](https://user-images.githubusercontent.com/23421917/201785260-a351906d-b311-4936-865e-ba5ed0af1cef.png)

This was decoded from a short sample @DJRaov recorded back 1-2 years ago. I have pushed everything, so I would highly recommend attempting to received and harvesting CADU to figure out the rest. While it should just be J2K, I find the images a bit too dark.

**Comment by sam210723** [comment]
Are they using a weird bit-depth for the images? I'll see what I can setup here for long CADU captures.

**Comment by Aang23** [comment]
In a way they are. They downlink it in 16-bits, but only the lower 8-bits contain any sort of useful imagery!

I think a small dish should be fine enough on this downlink.

**Comment by Foxiks** [comment]
![ht](https://user-images.githubusercontent.com/61653684/201976000-08ae8ec9-2cb8-4861-9f71-4d3d7908aaab.png)
This does not save data on Win and Kali

**Comment by Aang23** [comment]
Did the PID ever go green?

**Comment by Foxiks** [comment]
hm. no. Between 0 and ~8000

**Comment by Aang23** [comment]
How long did you wait for? CADUs are not always transmitted, waiting a while can be necessary until you see some.

**Comment by Foxiks** [comment]
~2 min 10 sec

**Comment by Aang23** [comment]
Yes, so wait at least 15 mins or such.

**Comment by Foxiks** [comment]
Okay!

**Comment by Aang23** [comment]
@Foxiks any luck?

**Comment by Foxiks** [comment]
I haven't tried it yet, but I should try later.

**Comment by eswnl** [comment]
Hi, just downloaded the latest 1.0.1 release.
When trying to decode FY-4A, it shows as GK-2A decoder at the top. Is this an error?
![image](https://user-images.githubusercontent.com/59115800/202675457-817c3294-ae2d-40be-9b22-70824b791696.png)

**Comment by Aang23** [comment]
Yes, that's a mistake on my end... Forgot to rename it. I also see it crashed on Windows, so another thing to look at.

**Comment by Aang23** [comment]
@eswnl just fixed this.

**Comment by eswnl** [comment]
I can now process the ts streams. Thanks Aang23.
![image](https://user-images.githubusercontent.com/59115800/202729281-f7909c6a-f9e7-4597-8645-578db2e14ca1.png)

Might be worth putting ts = true for HRIT as well.

**Comment by Aang23** [comment]
Indeed, let me do it. Though for HRIT we have not yet seen data in the clear.

**Comment by Aang23** [comment]
@eswnl just did it.
Is all your LRIT data looking so dark? (And did you manage to get a full disk yet perhaps?)

**Comment by eswnl** [comment]
Yes they are dark. Now testing LRIT, will this output images automatically when it detects the end? Also what is the freq option under "TS input"?
 
![image](https://user-images.githubusercontent.com/59115800/202837943-156a2788-965d-4bf7-91e9-9fe54fb5e9c5.png)

**Comment by eswnl** [comment]
Right, just processed the CADU file and got the image. I made the image brighter.
![image](https://user-images.githubusercontent.com/59115800/202839243-d0264035-9a1a-4007-a31d-2ead2bec988c.png)
![image](https://user-images.githubusercontent.com/59115800/202839254-05e2e4f3-d776-4b48-901b-1823474c6c94.png)

**Comment by eswnl** [comment]
Oh, just another thing, the program crashes when I click stop processing. Am using Airspy R2. But the CADU file is output correctly.

**Comment by eswnl** [comment]
HRIT demodulation was a bit unstable. Could not get any CADU files.
![image](https://user-images.githubusercontent.com/59115800/202841564-adbd0c12-e505-443f-9186-141e30c28f29.png)
Will continue to use the TBS5927 for HRIT.

**Comment by sam210723** [comment]
Nice work John, what is your RF setup for this?

**Comment by eswnl** [comment]
1.2m dish, Goes sawbird, patch feed, Airspy R2.
For HRIT, I have to use the TBS5927.

**Comment by eswnl** [comment]
Is it worth adding a decimation function? Airspy 10MSps seems a bit excessive for LRIT/HRIT, is it possible to gain some extra SNR?

**Comment by eswnl** [comment]
hmm, there is a ch1 visible on the LRIT at 10992x11200. But still using the low bit depth.
![image](https://user-images.githubusercontent.com/59115800/202881634-7f19fdae-2ae0-45f5-aed7-4cded96d9b83.png)

**Comment by sam210723** [comment]
How often are they sent on LRIT?

**Comment by eswnl** [comment]
Seems that ch1, ch9, ch12 are sent every 15min on the hour. Only ch1 is at 10992x11200.

**Comment by eswnl** [comment]
Seems HRIT signals are not constant, they come at xx:00 to xx:17 and xx:35 to xx:42.

**Comment by eswnl** [comment]
Ch12 and ch2 came in the clear on HRIT. But ch2 image was blank.
![FY4A_12_20221120T063400Z](https://user-images.githubusercontent.com/59115800/202889607-0079e429-7b3f-4b1b-817d-4c5a7d43a98d.png)
Looks like higher bit depth to me.

**Comment by Aang23** [comment]
@eswnl There are 2 different HRIT downlinks. The xx:00 and xx:35 ones are not the same, so that may be why SatDump's demodulator did not lock.

Indeed, that HRIT image is definitely higher bit depth. Do you have the data on hand? (TS/CADU), looks like I'll need to fixup some  stuff to output it correctly... Good to know HRIT *does* have some stuff in the clear!

**Comment by eswnl** [comment]
Ok, I'll send the TS streams. The TBS5927 did all the demodulation, not SatDump.
Another error, FY-4A HRIT decoder shows up as LRIT in the UI.

**Comment by Aang23** [comment]
![FY4A_12_20221120T061400Z](https://user-images.githubusercontent.com/23421917/202914168-270851f0-c7eb-48a6-b7b2-cca616849e7b.png)

Indeed, HRIT contains full 10-bits imagery. LRIT appears to only be 4-bits!? I'm thinking this may be somewhat due to JPEG-2000, maybe I need to increase bit depth *before* the inverse wavelet transform...

In HRIT, I have also found several "fast scan" streams. I am probably not decoding them properly yet, but hard to reverse-engineer without... More data. I'm now dumping them to imagery though. They appear to be called "REGC / REGX / RUEC".

![FY4A_12_20221120T063400Z](https://user-images.githubusercontent.com/23421917/202914280-6c053a7d-9e14-4404-9b6b-20f72164cf24.png)


> Ok, I'll send the TS streams. The TBS5927 did all the demodulation, not SatDump. I only fed SatDump the ts. Another error, FY-4A HRIT decoder shows up as LRIT in the UI.
Actually this it not a mistake. LRIT/HRIT are basically identical, apart from the contents. Hence, the same decoder is used and works exactly the same way (apart from different bit depths, etc).

As a note, I have just pushed the code that processes HRIT and the other imagery correctly.

**Comment by eswnl** [comment]
Thanks @Aang23 

REGC = China Regional
REGX = Regional
RUEC = ??

I'll see if I can get more HRIT data and any on 1681 MHz, although I've never seen a signal there before.

**Comment by Aang23** [comment]
I've got no clue about RUEC honestly. There's pretty much no documentation on FY-4 xRIT on this side of the firewall... I've only worked off the (very) scarce docs from @sam210723 and reverse-engineering.

There are 3 HRIT "types", HRIT-II and III are the ones you have decoded as of now... So we still lack HRIT-I, which per some old documentation is said to be on 1681. But of course it's hard to be sure.

**Comment by eswnl** [comment]
I checked to see any signal on 1681 MHz. I don't think it exists.
I also can't find any frequency listings for FY-4B, my feeling is that it doesn't use L-band.

**Comment by Aang23** [comment]
Interesting. Then I wonder where HRIT-I has gone!
As far as I could find 4B should also have xRIT on L-Band, but who knows... There's so little documentation.

Did you try pointing at it?

**Comment by eswnl** [comment]
Actually, I concluded that since HRIT-1 would overlap with HRIT 1679 MHz, I figured that it might be on a different polarisation. So I swapped the polarisation and found it at 1680 MHz. Wider signal and too weak to do anything with (for me).
![image](https://user-images.githubusercontent.com/59115800/204120516-bd261243-c784-400c-9f37-da7db015c468.png)
I think it uses much higher data rate.

**Comment by eswnl** [comment]
FY-4B is near GK-2A so somehow their frequencies would have to coexist as well.

**Comment by Aang23** [comment]
Ok, that's definitely it. HRIT-I likely contains all instruments / channels in full resolution I'd assume. I've got no idea if it contains anything in the clear but that's always a possibility...

Perhaps L-Band is currently off on 4B? I've looked at some pictures of the S/C and it definitely appears to carry the same L-Band antennas as 4A.

**Comment by Aang23** [comment]
Anyway, to do a little summary : 
- LRIT is fully supported and contains 3 channels every 15 mins. Always in the clear. Bit depth appears to be around 4-bits
- HRIT-II and HRIT-III contain *some* unencrypted channels at 10-bits (apparently, including one other VIS compared to LRIT, perhaps could allow composites?). Full disks are fully supported, partial scans still need to be figured out properly.
- HRIT-I does exist, but un-decoded at the moment.

**Comment by Aang23** [comment]
I've now made them able to live-process to imagery.
![Screenshot from 2022-11-27 14-17-32](https://user-images.githubusercontent.com/23421917/204137376-a692feaf-c694-4002-8d14-349c79d7a526.png)

**Comment by eswnl** [comment]
Tried pointing at FY-4B, all I found was this:
![image](https://user-images.githubusercontent.com/59115800/204255883-8a8004da-971d-478a-95a4-1ec9fa11fe93.png)

**Comment by Aang23** [comment]
Interesting. Though that's more likely FY-2x telemetry or such.

**Comment by sergsero** [comment]
Hello,

I am trying to make a long session of receiving HRIT signals FY4A at the frequency of 1679 MHz. Unfortunately, after switching the transmission from a broadband signal (~6 MHz) to a narrowband signal (~1.5 MHz), there is no repeated automatic recognition of the narrowband signal by the demodulator.

![20230506_1_FY4A_HRIT-32-45_1679MHz](https://user-images.githubusercontent.com/4005442/236617443-bee8f67b-0f65-42be-95b5-a8032b71295f.png)

![20230506_2_FY4A_HRIT-42-xx_1679MHz](https://user-images.githubusercontent.com/4005442/236617455-2450f48d-8196-4139-aea1-96c0a7b3afe7.png)

![20230506_3_FY4A_HRIT-00-xx_1679MHz](https://user-images.githubusercontent.com/4005442/236617463-794e12cb-a600-4a03-bf40-d98d54e4950b.png)

You have to manually perform the stop-start procedure with the demodulator button.

![20230506_FY4A_HRIT-00-xx_1679MHz-1](https://user-images.githubusercontent.com/4005442/236617575-4bddb55f-e6ed-4eae-8be3-b356264e26d3.png)

Is there any way to solve this problem?


Interestingly, sometimes the transmission of the image is interrupted in the current narrowband session by switching to broadband. And the missing part of the image is transmitted already in the next time narrowband mode.
Now, due to the forced manual break (stop/start demodulator) of the session, it is not possible to get a whole image.

![20230506_FY4A_HRIT-00-xx_1679MHz-3](https://user-images.githubusercontent.com/4005442/236617517-ffacb8d7-1892-402b-be3c-df0c2f4f75af.png)

![20230506_FY4A_HRIT-00-xx_1679MHz-4](https://user-images.githubusercontent.com/4005442/236617525-940dc68a-9e54-4f0a-9cf6-9baba79b1f9e.png)

**Comment by pluggerlockett** [comment]
Morning all. @eswnl @sergsero can you please provide me some information regarding transmission times for FY-4A HRIT? I'd like to take another swing at it. Historically I've been able to get sync with my 1.2m offset but I never received any data. I'm also happy to point at FY-4B for testing too, both are easily visible from mine.

**Comment by Aang23** [comment]
@pluggerlockett @eswnl @sam210723 did any of you ever collect more data from FY-4A to finish things up? And FY-4B to add pipelines? I wouldn't mind closing this if it can be considered done.

**Comment by pluggerlockett** [comment]
I looked at both FY-4A and FY-4B on Saturday evening with my 1.2m offset to attempt to add more input @Aang23. FY-4A seems to be operating nominally. LRIT is still on the same schedule, and HRIT was observed transmitting in the following timeslots

- 00:00 to 00:16

- 00:32 to 00:40

Sadly I didn't receive any unencrypted data despite having what I would expect to be a good enough SNR.

![FY-4A-HRIT](https://github.com/SatDump/SatDump/assets/56950661/2a6af4d9-f952-4876-b570-696079acc69b)

I then attempted to look at FY-4B with less than stellar results. The LRIT signal has changed significantly, last time I looked they had pulled back the power output to not interfere with GK-2A HRIT. But on Saturday it was basically a stonking huge carrier and not much else.

![FY-4B-LRIT](https://github.com/SatDump/SatDump/assets/56950661/66d80bbd-67fd-47ea-a86b-766a1e10737b)

Honestly, I'm not sure what to make of that. I then attempted to look at FY-4B HRIT, in nearly 30 minutes I didn't see anything transmit, and I was looking from ~00:20 to 00:48.

I hope that helps?

**Comment by Mifaxav** [comment]
I can receive FY-4A's LRIT signal from India here (no decode yet). 

Not the best antenna (helix) and I haven't received anything from FY-4 yet, but willing to help if needed.

**Comment by Mifaxav** [comment]
Forgot to add this

![image](https://github.com/user-attachments/assets/6946ecc2-0568-465e-a04f-019850593d98)

**Comment by kisaa** [comment]
My tests have shown that there is currently an HRIT (-II ?) signal from FY-4A (1697 MHz, BPSK ! modulation, 1368 ksps) and an LRIT signal from FY-4B (1697 MHz, DVB/S2, 150 kbit, modcod 9). FY-4B LRIT still only provides encrypted images. And I cannot decode the FY-4A HRIT signal using the tools available in SatDump: it does not look like DVB-S (no RS-Viterbi synchronization), nor DVB-S2 (it is BPSK-modulated), nor the CCSDS standard (also no Viterbi synchronization). Does anyone have any data on these signals?
